app/forms/date_form.py

```diff
@@ -16,191 +16,125 @@
+                    'year': substrings[0],
+                    'month': substrings[1],
+                    'day': substrings[2],
+                }
+            if len(substrings) == 2:
+                data = {'year': substrings[0], 'month': substrings[1]}
+            if len(substrings) == 1:
+                data = {'year': substrings[0]}
 
         super().process(formdata, data)
 
 
-def get_date_form(answer_store, metadata, answer=None, error_messages=None):
-    """
-    Returns a date form metaclass with appropriate validators. Used in both date and
-    date range form creation.
-
-    :param error_messages: The messages during validation
-    :param answer: The answer on which to base this form
-    :param answer_store: The current answer store
-    :param metadata: metadata for reference meta dates
-    :return: The generated DateForm metaclass
-    """
-
-    class DateForm(Form):
-        day = StringField()
-        year = StringField()
-        month = StringField()
-
-        @property
-        def data(self):
-            data = super().data
-            if not data['year'] and not data['month'] and not data['day']:
-                return None
-
-            return '{:04d}-{:02d}-{:02d}'.format(
-                int(data['year']), int(data['month']), int(data['day'])
-            )
-
-    validate_with = [OptionalForm()]
-
-    if answer['mandatory'] is True:
-        validate_with = validate_mandatory_date(error_messages, answer)
-
-    error_message = (
-        get_bespoke_message(answer, 'INVALID_DATE') or error_messages['INVALID_DATE']
-    )
-
-    validate_with.append(DateCheck(error_message))
+class CachedProperty:
+    """ A property that is only computed once per instance and then replaces
+        itself with an ordinary attribute. Deleting the attribute resets the
+        property.
 
-    if 'minimum' in answer or 'maximum' in answer:
-        min_max_validation = validate_min_max_date(
-            answer, answer_store, metadata, 'yyyy-mm-dd'
-        )
-        validate_with.append(min_max_validation)
+        Source: https://github.com/bottlepy/bottle/commit/fa7733e075da0d790d809aa3d2f53071897e6f76
+        """
 
-    DateForm.month = StringField(validators=validate_with)
+    def __init__(self, func):
+        self.__doc__ = getattr(func, '__doc__')
+        self.func = func
 
-    return DateForm
+    def __get__(self, obj, cls):
+        if obj is None:
+            return self
+        value = obj.__dict__[self.func.__name__] = self.func(obj)
+        return value
 
 
-def get_month_year_form(answer, answer_store, metadata, error_messages):
-    """
-    Returns a month year form metaclass with appropriate validators. Used in both date and
-    date range form creation.
+class DateForm(Form):
+    @CachedProperty
+    def data(self):
+        data = super().data
+        try:
+            if all(k in data for k in ('day', 'month', 'year')):
+                return '{year:04d}-{month:02d}-{day:02d}'.format(
```

Comment:
Might be a bit cleaner:
                `return f'{data['year']:04d}-{data['month']:02d}-{data['day']:02d}'`

Reply:
would it not have to be 
`return f'{int(data['year']):04d}-{int(data['month']):02d}-{int(data['day']):02d}'`

Reply:
Yeah, you're right. I guess we could cast those in one place, but doesn't make much difference overall.